### PR TITLE
[graph_trainer] WGrad fusion with grad accumulation

### PR DESCRIPTION
**Pull request by IvanKobzarev**
## Stack

Depends on #4666 (`graph-trainer-unified-pipeline-runtime`). Review only the top commit:

- `fc0759963 [graph_trainer] Accumulate WGrads in stage graphs`

## Why

GraphPP currently returns every microbatch's raw parameter gradients to Python and accumulates them there. This keeps a materialized WGrad tensor alive next to its accumulator and prevents a compatible producer from writing directly into persistent gradient state.

## Design

- Allocate one persistent raw-gradient accumulator set per local stage.
- Zero it when the runtime prepares a schedule invocation.
- Bind it as an explicit input to the same backward graph for every microbatch.
- Preserve the existing schedule and deferred-FSDP `REDUCE_GRAD` boundary.
- Preserve duplicate gradient slots and non-tensor DTensor metadata in the flat graph ABI.
- Expose graph-owned buffers directly as `param.grad`, without cloning.
- Treat committing that same tensor view again as a no-op, preventing CUDA-graph replay from adding an accumulator to itself.

Two independent configuration controls make the policy explicit:

- `gradient_accumulation_mode = auto | runtime | in_graph`
- `gradient_accum_in_wgrad_fusion = auto | disabled | enabled`

Explicit WGrad fusion also selects in-graph accumulation. In-graph accumulation can be selected independently while fusion remains disabled. Auto mode preserves the existing topology-based accumulation choice and gates fusion on `numerics_changing_optim`.

Gradient identity uses GraphTrainer's existing annotation context. A temporary `aten.alias` associates each autograd result with its parameter FQN; mandatory normalization transfers that metadata to the producer and removes the alias before partitioning.

The fusion pass has one narrow contract:

```
annotated add_(accumulator, wgrad_producer) -> accumulating producer
```

It performs no upstream FX traversal. The producer must directly feed the accumulation sink and have no other users. A successful rule retargets the existing producer in place and removes only the redundant `add_`:

- BF16 `aten.mm` -> `aten.addmm_`
- MXFP8 `aten._scaled_mm` -> `aten._scaled_addmm_`, when that operator is available

Unsupported layouts and producers retain the explicit in-graph `add_`. Bucketed FSDP currently also uses this correct fallback because bucket packing separates the logical WGrad producer from the accumulation sink; producer fusion before bucket packing is a follow-up.

## Validation

- Focused unit/runtime suite: 105 passed, 3 skipped.
- The BF16 unit test executes the rewritten `addmm_` graph and verifies that it updates and returns the persistent accumulator.
- A real traced BF16 `nn.Linear` stage verifies that parameter-gradient annotation drives direct producer fusion.
- PP=1, MB=1, explicit in-graph accumulation with fusion disabled vs. the joint graph:
  - bitwise-identical loss and grad norm for four steps
- PP=1, GA=4, explicit runtime accumulation vs. automatic in-graph accumulation:
  - bitwise-identical loss and grad norm for four steps
- PP=1, FSDP=2, GA=4, reduction inside backward, explicit in-graph vs. runtime accumulation:
  - bitwise-identical loss and grad norm for four steps
- PP=1, no FSDP, GA=4, CUDA graphs off vs. on:
  - bitwise-identical loss and grad norm for four steps
- PP=1, FSDP=2, GA=4, deferred FSDP reduction, CUDA graphs off vs. on:
  - bitwise-identical loss and grad norm for four steps
- `pre-commit`: all hooks pass except Pyrefly, which reports 12 existing errors in unrelated repository files.

## Current scope

The core Trainer captures the complete graph-runtime step, so the persistent stage buffers are exercised under CUDA-graph replay. DistMoE producer fusion remains a follow-up until the mutable backend operators are available in this checkout.

Real PP with multiple outer gradient-accumulation groups still needs an optimizer-step-aware graph-buffer reset boundary. Until that is added, runtime accumulation is the safe mode for that configuration.

### CODE REVIEW

File: torchtitan/experiments/graph_trainer/common_utils.py [hunk 1]
```diff
@@ -141,6 +141,26 @@ def _maybe_materialize_grad_for_param_layout(
     return materialized_grad
 
 
+def _is_same_tensor_view(lhs: torch.Tensor, rhs: torch.Tensor) -> bool:
+    if lhs is rhs:
+        return True
+    if isinstance(lhs, DTensor) != isinstance(rhs, DTensor):
+        return False
+    if isinstance(lhs, DTensor):
+        assert isinstance(rhs, DTensor)
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/common_utils.py]: nit: Not required given the conditions above.

File: torchtitan/experiments/graph_trainer/configs.py [hunk 2]
```diff
@@ -112,6 +112,25 @@ class GraphTrainerCompileConfig(CompileConfig):
       then reduces them with one ``REDUCE_GRAD`` action per stage.
     """
 
+    gradient_accumulation_mode: Literal["auto", "runtime", "in_graph"] = "auto"
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/configs.py:119]: If `in-graph` is feasible, then why would someone opt for `runtime` ? 
Do we really need this config option now or can we add it later when required?

File: torchtitan/experiments/graph_trainer/configs.py [hunk 3]
```diff
@@ -112,6 +112,25 @@ class GraphTrainerCompileConfig(CompileConfig):
...
+    """Choose where gradients accumulate across schedule microbatches.
+
+    - ``auto`` uses in-graph accumulation when the schedule has multiple
+      microbatches and gradients are local or FSDP reduction is deferred.
+    - ``runtime`` returns each microbatch's gradients to ``GraphRuntime``.
+    - ``in_graph`` accumulates into persistent graph inputs.
+    """
+
+    gradient_accum_in_wgrad_fusion: Literal["auto", "disabled", "enabled"] = "auto"
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/configs.py:132]: Same question for this -- can we add this config later when required?

File: torchtitan/experiments/graph_trainer/graph_pp/graph_builder.py [hunk 4]
```diff
@@ -910,6 +951,13 @@ def _apply_graph_pp_pre_partition_passes(
     )
 
     if not compile_config.enable_passes:
+        traced.gm = apply_graph_passes(
+            traced.gm,
+            traced.example_inputs,
+            construct_mandatory_graph_passes(),
+            compile_config=compile_config,
+            respect_disable_passes=False,
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/graph_pp/graph_builder.py:974]: Not introduced by this PR, but worth cleaning up: It is confusing to reason about `respect_disable_passes`. The passes construction should be deferred to constructors like construct_mandatory_graph_passes and this just applies the pass pipeline. 

For instance, in this case, we can have a construct_graph_pp_pre_partition_passes that does the right thing and we simplify apply_graph_passes

File: torchtitan/experiments/graph_trainer/graph_pp/runner.py [hunk 5]
```diff
@@ -450,7 +450,8 @@ def _populate_stage_states(self, stage: GraphPipelineStage) -> None:
         stage.state.flat_buffer_values = flat_buffer_values
         stage.state.trainable_params = trainable_params
         stage.state.unsharded_param_values = []
-        stage.state.unsharded_param_grads = []
+        graphs = self.stage_graphs[stage.stage_index]
+        stage.state.unsharded_param_grads = graphs.zero_grad_()
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/graph_pp/runner.py:465]: Hmm.. what if we have PP-microbatching and outer gradient accumulation in the trainer? Zeroing the grad buffers here will be a bug as GraphTrainer step might be invoked multiple times before optimizer.step.

We should either not do in-graph for this case, or clone param.grad in the GraphRuntime

File: torchtitan/experiments/graph_trainer/wgrad_accumulation.py [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+    producer.meta["original_aten"] = scaled_addmm
+    _replace_sink_with_producer(sink, producer)
+    return True
+
+
+def fuse_wgrad_accumulation_pass(
+    gm: fx.GraphModule,
+    example_inputs: tuple[Any, ...] | None = None,
+) -> fx.GraphModule:
+    """Fuse annotated BF16 dense WGrad accumulation when a kernel is available."""
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/wgrad_accumulation.py]: This docstring is incomplete as we also support MXFP8 sink in this file

**Review by aditvenk — CHANGES_REQUESTED** [review]
LGTM, except for the potential bug around PP + GA

File: torchtitan/experiments/graph_trainer/configs.py (hunk 2 above)
Comment (IvanKobzarev) [reply torchtitan/experiments/graph_trainer/configs.py:119]: This is a good question. My logic was - that if someone wants custom PP schedule and to have them in runtime.

File: torchtitan/experiments/graph_trainer/configs.py (hunk 3 above)
Comment (IvanKobzarev) [reply torchtitan/experiments/graph_trainer/configs.py:132]: I think the fusion could change numerics in comparison with non fused wgrad. That is why I wanted to give this option.

File: torchtitan/experiments/graph_trainer/graph_pp/graph_builder.py (hunk 4 above)
Comment (IvanKobzarev) [reply torchtitan/experiments/graph_trainer/graph_pp/graph_builder.py:974]: Agree, we should remove respect_disable_passes=False and have one place where we construct the passes.

File: torchtitan/experiments/graph_trainer/graph_pp/runner.py (hunk 5 above)
Comment (IvanKobzarev) [reply torchtitan/experiments/graph_trainer/graph_pp/runner.py:465]: Oh, yes.
I think will be good to extract ZERO_GRAD_ACCUMS into a special schedule action, that will be added for now only for PP=1, while using Runtime gradient accumulation for PP > 1 as it was before.

At the moment trainer does optimizer.zero_grad(), but it will be good to add this to the schedule too, to have full representation and control

Comment (aditvenk) [reply torchtitan/experiments/graph_trainer/graph_pp/runner.py:465]: Can we loudly error out if someone tries to do PP + in-graph GA?

**Review by aditvenk — APPROVED** [review]
